### PR TITLE
Packing and Unpacking for GEM and ME0

### PR DESCRIPTION
**Pull request by jshlee**
For slice test and full sim unpacking and packing is required.
Packer - GEMDigiToRawModule
Unpacker - GEMRawToDigiModule
DQM/GEM
CondCore/GEMPlugins
CondFormats/GEMObjects
CondFormats/DataRecord
CondTools/GEM
EventFilter/GEMRawToDigi

Talks
https://indico.cern.ch/event/671987/contributions/2753392/attachments/1539931/2414350/gem_20171012.pdf
https://indico.cern.ch/event/678737/contributions/2780039/attachments/1555935/2446778/gem_20171109.pdf
https://indico.cern.ch/event/675328/contributions/2763616/attachments/1547638/2430150/171024_Hyunyong_Run304810_TK_v4c.pdf

me0 packing and unpacking not yet in std seq due to me0 using pesudo digitiser
need to update configs with channel mapping from DB, currently using dummy mapping

@slava77 @kpedro88 @hyunyong

### CODE REVIEW

**Comment by kpedro88** [comment]
@jshlee please apply code checks (required before PR tests can proceed)

You should also prepare a PR for https://github.com/cms-sw/cms-bot to assign a category for the new package(s).

This is quite a lot of new code, so the review may take some time.

**Comment by perrotta** [comment]
please test

**Comment by kpedro88** [comment]
please test

File: EventFilter/GEMRawToDigi/plugins/GEMDigiToRawModule.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+#include "EventFilter/GEMRawToDigi/interface/AMC13Event.h"
+#include "CondFormats/DataRecord/interface/GEMEMapRcd.h"
+#include "CondFormats/GEMObjects/interface/GEMEMap.h"
+#include "CondFormats/GEMObjects/interface/GEMROmap.h"
+
+namespace edm {
+  class ConfigurationDescriptions;
+}
+
+class GEMDigiToRawModule : public edm::EDProducer {
```
Comment (slava77) [inline EventFilter/GEMRawToDigi/plugins/GEMDigiToRawModule.h]: legacy modules are not allowed

Comment (slava77) [reply EventFilter/GEMRawToDigi/plugins/GEMDigiToRawModule.h]: this should be of a type known to threaded framework
https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkModuleTypes
edm::global would be a better choice if implementation allows, edm::stream is OK.
edm::one is rather an exception and will need some justification.

Comment (jshlee) [reply EventFilter/GEMRawToDigi/plugins/GEMDigiToRawModule.h]: used global, but had to use doBeginRun_ to get up mapping. Is this correct?

**Comment by davidlange6** [comment]
hi @jshlee all - I see ~5 separate pull requests lumped into one request here. As the process of big PRs can take a while to converge, could I suggest we break this into pieces? To start with, PRs on CondFormats and DataFormats can process separately..

**Comment by jshlee** [comment]
@davidlange6 - yes I will break it up. 
I will keep this open for the final unpacker.

**Comment by slava77** [comment]
On 11/14/17 7:21 PM, Jason Lee wrote:
> @davidlange6 <https://github.com/davidlange6> - yes I will break it up.
> I will keep this open for the final unpacker.
> 

it may be difficult to keep this open and still consistent if the other
two PRs require updates.

**Comment by perrotta** [comment]
@jshlee : what is the status/plan for this PR? Are the sibling requests #21305 and #21306 ready to be merged in the master? If so could you please push for getting all signatures and have them signed? Once that done, I expect that you will rebase this PR to the IB which contains them, apply the possible needed updates, and then we can start evaluating it.

About the plans: is this PR only expected to be applied on 2018 data/MC, and therefore it can remain in the master only? Or do you plan to have it also in the re-reco of 2017 data/MC, and therefore a backport will also get eventually requested? Please let us know. possibly in time to discuss it at the next ORP (Tue Nov 28, I guess). Especially in case you plan any backport, some speed up on this and associated PRs would be welcome.

**Comment by slava77** [comment]
On 11/24/17 3:22 AM, perrotta wrote:
> Or do you plan to have it also in the re-reco of 2017 data/MC, and
> therefore a backport will also get eventually requested?

If a backport is needed for 94X, it will have to be made in a way that
does not modify the default processing workflow.
The 94X is the production release now and no-change policy applies.

**Comment by davidlange6** [comment]
Your PR is unmergeable. Please have a look and possibly rebase it.

**Comment by jshlee** [comment]
Yes, i will rebase and only add in the packing and unpacking here.
DQM stuff will come in another PR.

**Comment by perrotta** [comment]
@jshlee : as of now, both #21305 and #21306 have been merged in the master.
As soon as an IB will become available containing both those pull requests (let say, some time tomorrow) could you please rebase to that IB, so that the modifications already merged with them will not appear here any more?

**Comment by jshlee** [comment]
yep sure

**Comment by jshlee** [comment]
I rebased but it doesnt seem to update it here correctly.

comparing it here 
https://github.com/cms-sw/cmssw/compare/master...jshlee:GEMRawToDigi?expand=1
shows the correct changes

is there a way to fix this or should i just do a new PR?

**Comment by kpedro88** [comment]
Well, you didn't rebase, you merged. Probably easiest just to make a new PR.

**Comment by jshlee** [comment]
ok will just close this and open a new one.

**Pull request closed without merging**